gravitational/downtime
# integrate contentful mvp

- Addresses #69 

## Integrate Contentful CMS
- connect [this Vercel branch](https://downtime-git-delia-contentfulintegrate-contentful-gravitational.vercel.app/) with Contentful, allowing any updates to content on Contentful to trigger automatic re-deploys of the connected branch 
- use [`order`](https://www.contentful.com/developers/docs/references/content-delivery-api/#/reference/search-parameters/order) option in `getEntries` query to order posts by user-supplied "pubDate" field.
- generate RSS feed from `getStaticProps` function - this allows RSS feed to be updated anytime new content is published

## Contentful:
**Joke Model Fields**
<img width="858" alt="Screen Shot 2022-02-14 at 10 23 49 PM" src="https://user-images.githubusercontent.com/70108137/153986759-196e7dc2-c616-4490-8142-63fd5ebcb831.png">

**Current field notes**
- "anchor" uses a regex to accept only unique 5 digit numerical strings 

- "twitterHTMLEmbedFull" is an optional field to allow users to more easily parse the "twitterImage" from the embedded code they grab from twitter 

- added validation to "image" to allow the sizes shown. However, I had to keep increasing the range as I added additional images. This validation may need to be removed or images updated to fit within these parameters.
<img width="834" alt="Screen Shot 2022-02-14 at 9 37 30 PM" src="https://user-images.githubusercontent.com/70108137/153982003-a4c51ddd-4eff-445e-b9ee-d3123471cd19.png">

**Potential future improvements:**
- It might be challenging to keep track of the intended "anchor" value as the application grows - we could consider replacing this with the unique ID that Contentful generates for each entry or a unique slug:
  - Contentful offers a "slug" datatype, with the option to validate for uniqueness and restrict characters with a regex. We could easily replace "anchor" with a more human-friendly "slug" field

## Date Issue:
When "pubDate" is in UTC +00:00 timezone, the date and HTML title attribute date show different values, despite coming from the same data, as shown here:
<img width="644" alt="Screen Shot 2022-02-15 at 1 07 01 PM" src="https://user-images.githubusercontent.com/70108137/154122690-5bbe61cd-2708-4aa2-81b2-63f62856da3a.png">

One solution is to define the "pubDate" as UTC +05:00 (US EST) when creating the joke. I updated this to be the default for the "pubDate" field. It does not apply retroactively to jokes, so all jokes in Contentful will need to be verified if we want to use this solution.
 <img width="846" alt="Screen Shot 2022-02-15 at 1 09 10 PM" src="https://user-images.githubusercontent.com/70108137/154122800-65825d59-47c8-4ecd-bc6c-c24e47c7b490.png">
Jokes with a "pubDate" defined this way show matching values for date and HTML title attribute date as shown:
<img width="626" alt="Screen Shot 2022-02-15 at 1 13 43 PM" src="https://user-images.githubusercontent.com/70108137/154123581-a7f5689a-abc3-4731-8930-75691f134b1d.png">

 ## Test Integration in real time:
 1. Visit [this test branch](https://downtime-git-delia-contentfulintegrate-contentful-gravitational.vercel.app/)
 2. visit our Contentful account [here](https://app.contentful.com/spaces/9l2srfbba2yk/home). Publish one of the draft posts or create your own
 3. Return to the test branch once deploy finishes to see updates (refresh if needed).
